### PR TITLE
Make maxRetries configurable for endpoints and route-services

### PR DESCRIPTION
**Pull request by domdom82**
* A short explanation of the proposed change:
Fixes https://github.com/cloudfoundry/routing-release/issues/285

As [discussed on slack](https://cloudfoundry.slack.com/archives/C01ABMVNE9E/p1667563647259079) this PR makes the `MaxRetries` value of RequestHandler configurable and removes the constant value.

- The default will remain at 3 as this was hard-coded until now (this will be set [via a spec change](https://github.com/cloudfoundry/routing-release/pull/298) in routing-release)
- There will be a separate `MaxRetries` value for route-services. The reasoning behind this is we want the value of 0 to imply "retry indefinitely / until endpoints are exhausted", because we feel there is no great benefit in having a fixed value for retries, when the number of endpoints is dynamic.
- Route-services do not have a fixed number of endpoints, they are just a URL that is called. So here it would be dangerous to set a value of 0 for indefinite retries. If the route-service misbehaves and never responds, this would deadlock gorouter.
- Therefore, a second value was introduced, to retain the option to retry all possible endpoints while still keeping a hard-limit on route-service retries.

* An explanation of the use cases your change solves

1. When an app developer scales up their app to more than 3 instances, only 3 of them will ever be used in case of failing round trips. This reduces their app availability as increasing the number of instances does not yield a higher chance of success.
2. In outage scenarios where NATS may be overloaded, Diego cells have crashed or an AZ may be unstable, large portions of the route table can become stale. Even if the app developer has multiple app instances deployed, they may still receive a 502 error, if Gorouter happens to select only stale instances. Ironically, the more instances were deployed initially, hoping to increase availability, the less available the app will be, because Gorouter will have more stale endpoints to choose from, while still giving up after just 3 tries.

* Instructions to functionally test the behavior change using operator interfaces (BOSH manifest, logs, curl, and metrics)

1. Deploy routing-release that contains this change ([PR](https://github.com/cloudfoundry/routing-release/pull/298))
2. Change the `MaxRetries` property as needed (e.g. set it to 5 instead of 3)
3. Deploy a CF app with one instance
4. Simulate one of the above scenarios by adding stale endpoints to the app using the `nats-client`.
5. Save the route table using
```
/var/vcap/jobs/gorouter/bin/nats_client save routes.json
```
6. Modify the routes.json table by adding multiple stale endpoints e.g.
```
 { 
 "my-app.cf.com": [
// The first entry is the valid endpoint...
    {
      "address": "10.0.73.1:21003",
      "private_instance_id": "5d440ada-375a-4ac6-578d-713f",
      "protocol": "http2",
      "server_cert_domain_san": "5d440ada-375a-4ac6-578d-713f",
      "tags": {
        "app_id": "3c32ce67-fdd2-4159-88a7-86011250475a",
        "app_name": "my-app",
        "component": "route-emitter",
        "instance_id": "0",
        "organization_id": "bc578e33-bce1-4bce-91c3-a302bcb27ee3",
        "organization_name": "myorg",
        "process_id": "0695920e-d1ea-4a28-91d5-4fd62cff5d97",
        "process_instance_id": "5d440ada-375a-4ac6-578d-713f",
        "process_type": "web",
        "source_id": "3c32ce67-fdd2-4159-88a7-86011250475a",
        "space_id": "442f1c44-3f4b-4db8-a964-4a74d8e044bf",
        "space_name": "dev"
      },
      "tls": true,
      "ttl": 120
    },
// ... followed by a list of stale entries... ( > 3)
    {
      "address": "10.99.99.1:21003",
      "private_instance_id": "5d440ada-375a-4ac6-578d-713f",
      "protocol": "http2",
      "server_cert_domain_san": "5d440ada-375a-4ac6-578d-713f",
      "tags": {
        "app_id": "3c32ce67-fdd2-4159-88a7-86011250475a",
        "app_name": "my-app",
        "component": "route-emitter",
        "instance_id": "1",
        "organization_id": "bc578e33-bce1-4bce-91c3-a302bcb27ee3",
        "organization_name": "myorg",
        "process_id": "0695920e-d1ea-4a28-91d5-4fd62cff5d97",
        "process_instance_id": "5d440ada-375a-4ac6-578d-713f",
        "process_type": "web",
        "source_id": "3c32ce67-fdd2-4159-88a7-86011250475a",
        "space_id": "442f1c44-3f4b-4db8-a964-4a74d8e044bf",
        "space_name": "dev"
      },
      "tls": true,
      "ttl": 120
    },
    {
      "address": "10.99.99.2:21003",
      "private_instance_id": "5d440ada-375a-4ac6-578d-713f",
      "protocol": "http2",
      "server_cert_domain_san": "5d440ada-375a-4ac6-578d-713f",
      "tags": {
        "app_id": "3c32ce67-fdd2-4159-88a7-86011250475a",
        "app_name": "my-app",
        "component": "route-emitter",
        "instance_id": "1",
        "organization_id": "bc578e33-bce1-4bce-91c3-a302bcb27ee3",
        "organization_name": "myorg",
        "process_id": "0695920e-d1ea-4a28-91d5-4fd62cff5d97",
        "process_instance_id": "5d440ada-375a-4ac6-578d-713f",
        "process_type": "web",
        "source_id": "3c32ce67-fdd2-4159-88a7-86011250475a",
        "space_id": "442f1c44-3f4b-4db8-a964-4a74d8e044bf",
        "space_name": "dev"
      },
      "tls": true,
      "ttl": 120
    },
(...)
  ]
}
```
7. Load the route table back into gorouter
```
/var/vcap/jobs/gorouter/bin/nats_client load routes.json
```
8. Curl the app locally on the gorouter
```
curl -v https://localhost:443/ -H "host: my-app.cf.com"
```

* Expected result after the change

200 OK (from the app)

* Current result before the change

502 Bad Gateway: Registered endpoint failed to handle the request. (from gorouter)

* Links to any other associated PRs

* [x] I have viewed signed and have submitted the Contributor License Agreement

* [x] I have made this pull request to the `main` branch

* [x] I have run all the unit tests using `scripts/run-unit-tests-in-docker` from [routing-release](https://github.com/cloudfoundry/routing-release).

* [ ] (Optional) I have run Routing Acceptance Tests and Routing Smoke Tests on bosh lite

* [ ] (Optional) I have run CF Acceptance Tests on bosh lite

### CODE REVIEW

**Comment by ameowlia** [comment]
@nicregez, can you please rebase develop (for routing-release) and main (for gorouter) into these PR branches so I can more easily test it with the current cf-deployment? Thanks!

**Comment by nicregez** [comment]
@ameowlia Thanks for having taken up our request so quickly. I merged both branches as you requested

**Comment by ameowlia** [comment]
Hi @nicregez , I was testing this out and I found a small bug. 

🐛 `num-endpoints` in the logs does not update after failed endpoints are pruned.

### Testing Setup
* My CF has 1 router, 2 diego cells, and 1 nats VMs
* I deployed with this code and set max_attempts to 5.
* I pushed 6 instances of an app.
* I added iptables rules to drop all traffic to the diego cells and to drop all traffic to nats (to prevent the routes from re-registering)
   ```
   $ iptables -A OUTPUT -d 10.0.1.14 -j DROP # diego cell 1
   $ iptables -A OUTPUT -d 10.0.1.13 -j DROP # diego cell 2
   $ iptables -A OUTPUT -d 10.0.1.0 -j DROP # nats 
   ```
* In one terminal, I sshed onto the router VM and watched the count 
   ```
   $ cd /var/vcap/jobs/gorouter/bin
   $ watch -n 0.5 "./retrieve-local-routes | jq '.\"potato.spanishpink.cf-app.com\"' | jq length"
   ```
* In a 2nd terminal, I sshed onto the router VM and tailed the logs
   ```
   $ tail -f gorouter.stdout.log | grep backend-endpoint-failed
   ```
* In a 3rd terminal, I curled the app.

### What I saw

1. The endpoints were pruned correctly
2. However the logs did not reflect the new endpoint count.

**Log Output**
`num-endpoints` is all the way at the end of the log line.
```
{"log_level":3,"timestamp":"2022-12-16T16:11:32.125521238Z","message":"backend-endpoint-failed","source":"vcap.gorouter","data":{"route-endpoint":{"ApplicationId":"82b2fd9a-00f1-4c78-a1af-1001648da7d4","Addr":"10.0.1.14:61010","Tags":{"app_id":"82b2fd9a-00f1-4c78-a1af-1001648da7d4","app_name":"potato","component":"route-emitter","instance_id":"3","organization_id":"16077d3a-42b6-4594-9506-03c59435e5c7","organization_name":"o","process_id":"82b2fd9a-00f1-4c78-a1af-1001648da7d4","process_instance_id":"522ab663-bce3-431d-42d6-36a9","process_type":"web","source_id":"82b2fd9a-00f1-4c78-a1af-1001648da7d4","space_id":"05cb450a-9baf-4777-a84a-88ab65c4ce50","space_name":"s"},"RouteServiceUrl":""},"error":"dial tcp 10.0.1.14:61010: i/o timeout","attempt":1,"vcap_request_id":"f8273690-4962-4e3f-4b7b-c6b3cbcad907","retriable":true,"num-endpoints":6}}
{"log_level":3,"timestamp":"2022-12-16T16:11:37.126315987Z","message":"backend-endpoint-failed","source":"vcap.gorouter","data":{"route-endpoint":{"ApplicationId":"82b2fd9a-00f1-4c78-a1af-1001648da7d4","Addr":"10.0.1.13:61002","Tags":{"app_id":"82b2fd9a-00f1-4c78-a1af-1001648da7d4","app_name":"potato","component":"route-emitter","instance_id":"0","organization_id":"16077d3a-42b6-4594-9506-03c59435e5c7","organization_name":"o","process_id":"82b2fd9a-00f1-4c78-a1af-1001648da7d4","process_instance_id":"232663f9-ddf2-43a1-758d-4fb0","process_type":"web","source_id":"82b2fd9a-00f1-4c78-a1af-1001648da7d4","space_id":"05cb450a-9baf-4777-a84a-88ab65c4ce50","space_name":"s"},"RouteServiceUrl":""},"error":"dial tcp 10.0.1.13:61002: i/o timeout","attempt":2,"vcap_request_id":"f8273690-4962-4e3f-4b7b-c6b3cbcad907","retriable":true,"num-endpoints":6}}
{"log_level":3,"timestamp":"2022-12-16T16:11:42.126993224Z","message":"backend-endpoint-failed","source":"vcap.gorouter","data":{"route-endpoint":{"ApplicationId":"82b2fd9a-00f1-4c78-a1af-1001648da7d4","Addr":"10.0.1.14:61007","Tags":{"app_id":"82b2fd9a-00f1-4c78-a1af-1001648da7d4","app_name":"potato","component":"route-emitter","instance_id":"5","organization_id":"16077d3a-42b6-4594-9506-03c59435e5c7","organization_name":"o","process_id":"82b2fd9a-00f1-4c78-a1af-1001648da7d4","process_instance_id":"5087cf6c-6857-414f-6c5e-9a94","process_type":"web","source_id":"82b2fd9a-00f1-4c78-a1af-1001648da7d4","space_id":"05cb450a-9baf-4777-a84a-88ab65c4ce50","space_name":"s"},"RouteServiceUrl":""},"error":"dial tcp 10.0.1.14:61007: i/o timeout","attempt":3,"vcap_request_id":"f8273690-4962-4e3f-4b7b-c6b3cbcad907","retriable":true,"num-endpoints":6}}
{"log_level":3,"timestamp":"2022-12-16T16:11:47.127973523Z","message":"backend-endpoint-failed","source":"vcap.gorouter","data":{"route-endpoint":{"ApplicationId":"82b2fd9a-00f1-4c78-a1af-1001648da7d4","Addr":"10.0.1.13:61010","Tags":{"app_id":"82b2fd9a-00f1-4c78-a1af-1001648da7d4","app_name":"potato","component":"route-emitter","instance_id":"4","organization_id":"16077d3a-42b6-4594-9506-03c59435e5c7","organization_name":"o","process_id":"82b2fd9a-00f1-4c78-a1af-1001648da7d4","process_instance_id":"41863be4-646a-47ee-46ea-9240","process_type":"web","source_id":"82b2fd9a-00f1-4c78-a1af-1001648da7d4","space_id":"05cb450a-9baf-4777-a84a-88ab65c4ce50","space_name":"s"},"RouteServiceUrl":""},"error":"dial tcp 10.0.1.13:61010: i/o timeout","attempt":4,"vcap_request_id":"f8273690-4962-4e3f-4b7b-c6b3cbcad907","retriable":true,"num-endpoints":6}}
{"log_level":3,"timestamp":"2022-12-16T16:11:52.128949727Z","message":"backend-endpoint-failed","source":"vcap.gorouter","data":{"route-endpoint":{"ApplicationId":"82b2fd9a-00f1-4c78-a1af-1001648da7d4","Addr":"10.0.1.13:61006","Tags":{"app_id":"82b2fd9a-00f1-4c78-a1af-1001648da7d4","app_name":"potato","component":"route-emitter","instance_id":"2","organization_id":"16077d3a-42b6-4594-9506-03c59435e5c7","organization_name":"o","process_id":"82b2fd9a-00f1-4c78-a1af-1001648da7d4","process_instance_id":"4b192f8c-54f2-4dd1-491c-2d57","process_type":"web","source_id":"82b2fd9a-00f1-4c78-a1af-1001648da7d4","space_id":"05cb450a-9baf-4777-a84a-88ab65c4ce50","space_name":"s"},"RouteServiceUrl":""},"error":"dial tcp 10.0.1.13:61006: i/o timeout","attempt":5,"vcap_request_id":"f8273690-4962-4e3f-4b7b-c6b3cbcad907","retriable":true,"num-endpoints":6}}
```

Please fix this before we merge. Thanks!

**Comment by nicregez** [comment]
@ameowlia Thanks for your thorough testing and report. I pushed another commit and tested the logging, aligned with the bug you reported 2 days ago.

File: proxy/round_tripper/proxy_round_tripper.go [hunk 1]
```diff
@@ -170,6 +169,7 @@ func (rt *roundTripper) RoundTrip(originalRequest *http.Request) (*http.Response
 				iter.EndpointFailed(err)
 
 				retriable := rt.retriableClassifier.Classify(err)
+				numberOfEndpoints := reqInfo.RoutePool.NumEndpoints()
```
[the remark this answers is not in the document]
Comment (ameowlia) [reply proxy/round_tripper/proxy_round_tripper.go:172]: 🤔 Ok, you've convinced me @domdom82 .

😅 @nicregez , if you revert that last commit I will merge this in. Thank you for your patience and hard work!

**Pull request merged**